### ISSUE TITLE
Expanding the truncated commit messages... = Error: Bad credentials

### ISSUE THREAD

**Issue by Stericycle-MMorris** [rfennell/AzurePipelines#620]
### Azure DevOps Extension you are using
XplatGenerateReleaseNotes
 Generate Release Notes (Crossplatform) 

### Where are you running it?
- Azure DevOps Service (VSTS)
- GitHub

### Version of Extension/Task
Version: 2.20.1 (Current Public Release / Azure Market)

### Expected behaviour and actual behaviour
Expected behaviour: 

Pull commit info from GitHub
Generate release notes to Release.md

Actual behaviour: 

Getting the current build details
Expanding the truncated commit messages...
##[error]Error: Bad credentials
##[error]Error: Bad credentials
Finishing: XplatGenerateReleaseNotes

### Steps to reproduce the problem
- task: XplatGenerateReleaseNotes@2
  inputs:
    outputfile: 'Release.md'
    templateLocation: 'File'
    templatefile: 'Documentation\ReleaseNoteTemplate.md'
    delimiter: ':'
    fieldEquality: '='

Unknown, task worked until this morning then recent commits which involved added .xml files to repo the task failed with the bad credentials issue. Nothing particularly strange about the commits although there was a message body on one of them instead of just title information.

Other Azure DevOps projects have the same setup / credentials and are still working.

Following the debug error messsage it appears the issue is occurring inside 'export async function expandTruncatedCommitMessages' and probably the line 'await restClient.get<GitCommit>(change.location)' 

The rest of the build process is successfully pulling Github data and building, other projects are still working on creating release notes (and then additionally submitting to the wiki using the uploader).

Any thoughts on why this credentials error is occurring?

**Comment by rfennell** [comment]
There has been updates to these extension for a few weeks, and I have just re-run my integration test suite and they pass.

So it would appear to be some effect based the commits, that is strange.

Can you re-run with a variable system.debug=true and let me see log file (anonymised as yo wish)

**Comment by Stericycle-MMorris** [comment]
I can confirm that it is definitely commit based, I eventually disabled the release notes tasks to get a working build out and have since re-enabled it, it now works without those commits in its queue/backlog?

I have run a build of the commit that first failed the release notes task with the debug variable in place and (unfortunately?) it succeeded but Azure noted that there were 0 commits in the changes when the error builds have 2 commits or more (I'd had an earlier broken build before release notes task due to forgetting to push some files).

I'll attempt to get a replication tomorrow but if that doesn't work then I'll close the ticket.

**Comment by Stericycle-MMorris** [comment]
I _finally_ got a replication, I believe it's to do with the commit messages or perhaps having multiple queued commits. 

I went back to the commit before the issue branched and repeated my steps as one commit (minus commit names). This worked which was annoying.

I re-branched a second time to do them one at a time which also worked. So it seemed like the commit contents wasn't causing the error.

As a final test I decided to test that the commit title / messages was the issue by just modifying my readme with an extra space / letter for two commits then ran the pipeline on those two outstanding commits with their original commit title / message. This replicated the issue.

Below is the log:

##[debug]agent.TempDirectory=d:\a\_temp
##[debug]loading inputs and endpoints
##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
##[debug]loading INPUT_DELIMITER
##[debug]loading INPUT_EMPTYSETTEXT
##[debug]loading INPUT_FIELDEQUALITY
##[debug]loading INPUT_OUTPUTFILE
##[debug]loading INPUT_SHOWONLYPRIMARY
##[debug]loading INPUT_SORTWI
##[debug]loading INPUT_STOPONREDEPLOY
##[debug]loading INPUT_TEMPLATEFILE
##[debug]loading INPUT_TEMPLATELOCATION
##[debug]loading SECRET_AZUREDEVOPSPAT
##[debug]loading SECRET_SYSTEM_ACCESSTOKEN
##[debug]loaded 14
##[debug]Agent.ProxyUrl=undefined
##[debug]Agent.CAInfo=undefined
##[debug]Agent.ClientCert=undefined
##[debug]Agent.SkipCertValidation=undefined
##[debug]Starting Tag XplatGenerateReleaseNotes task
##[debug]System.TeamFoundationCollectionUri=https://redacted.visualstudio.com/
##[debug]System.TeamProject=project
##[debug]templateLocation=File
##[debug]templatefile=d:\a\1\s\Documentation\ReleaseNoteTemplate.md
##[debug]inlinetemplate=null
    ##[debug]outputfile=Release.md
    ##[debug]outputVariableName=null
    ##[debug]emptySetText=None
    ##[debug]delimiter=:
    ##[debug]showOnlyPrimary=False
    ##[debug]fieldEquality==
    ##[debug]stopOnRedeploy=False
    ##[debug]SortWi=False
    ##[debug]System.AccessToken=***
    ##[debug]Creating the credential handler
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    (node:2656) Warning: Use Cipheriv for counter mode of aes-256-ctr
    ##[debug]Release.ReleaseId=undefined
    Getting the current build details
    ##[debug]Build.BuildId=1169
    Expanding the truncated commit messages...
    ##[debug]Expanding commit [558083fbc71bd653079c8f447f1c7038aa81fe60]
    ##[error]Error: Bad credentials
    ##[debug]Processed: ##vso[task.issue type=error;]Error: Bad credentials
    ##[debug]SYSTEM.HOSTTYPE=build
    ##[debug]BUILD.DEFINITIONNAME=redacted.project
    ##[debug]BUILD.BUILDID=1169
    ##[debug]SYSTEM.TEAMFOUNDATIONSERVERURI=https://redacted.visualstudio.com/
    ##[debug]SYSTEM.TEAMPROJECT=project
    ##[debug]BUILD.BUILDID=1169
    ##[debug]TASK.DISPLAYNAME=XplatGenerateReleaseNotes
    ##[debug]SYSTEM.JOBID=12f1170f-54f2-53f3-20dd-22fc7dff55f9
    ##[debug]AGENT.VERSION=2.164.7
    ##[debug]Processed: ##vso[telemetry.publish area=XplatGenerateReleaseNotes;feature=reliability]{"hostType":"build","definitionName":"redacted.project","processId":"1169","processUrl":"https://redacted.visualstudio.com/project/_build?buildId=1169","taskDisplayName":"XplatGenerateReleaseNotes","jobid":"12f1170f-54f2-53f3-20dd-22fc7dff55f9","agentVersion":"2.164.7","version":{"Major":"2","Minor":"20","Patch":"1"},"issueType":"error","errorMessage":"{\"stack\":\"Error: Bad credentials\\n    at RestClient.<anonymous> (d:\\\\a\\\\_tasks\\\\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\\\\2.20.1\\\\node_modules\\\\typed-rest-client\\\\RestClient.js:173:28)\\n    at next (native)\\n    at fulfilled (d:\\\\a\\\\_tasks\\\\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\\\\2.20.1\\\\node_modules\\\\typed-rest-client\\\\RestClient.js:6:58)\\n    at process._tickCallback (internal/process/next_tick.js:109:7)\",\"message\":\"Bad credentials\"}"}
    ##[debug]task result: Failed
    ##[error]Error: Bad credentials
    ##[debug]Processed: ##vso[task.issue type=error;]Error: Bad credentials
    ##[debug]Processed: ##vso[task.complete result=Failed;]Error: Bad credentials
    Finishing: XplatGenerateReleaseNotes


The commit [558083fbc71bd653079c8f447f1c7038aa81fe60] it chocked on was just adding a space to the readme file so I don't believe it's content related therefore it is likely the title / message:

Added missing publish.xml files for DriveAuto
Resolving OPS-1975 incorrect DB name in DACPAC variables

I suspect it either doesn't like the dot in the title or the hyphen in the message

**Comment by rfennell** [comment]
That is strange indeed. It looks to be failing in [ReleaseNotesFunctions.ts](https://github.com/rfennell/AzurePipelines/blob/master/Extensions/XplatGenerateReleaseNotes/V2/ReleaseNotesFunctions.ts) around line 107. This when an extra call is made to unpack the full commit message as opposed to just the start if the string (80 char. as I remember)

Can see an obvious reason why this might be failing. 

Sorry I am a little unclear over your repo, is this right so I can try to debug?

1. Repo with a text file in it
1. Edit the text file, add a character - then commit with a message 'Added missing publish.xml files for DriveAuto'
1. Edit the text file, add another a character - then commit with a message Resolving OPS-1975 incorrect DB name in DACPAC variables'
1. push to Azure DevOps
1. Run a build that includes these two commits
1. Run the release with the release notes generator - get the error

That does not seem enough, but I will try to repo it

**Comment by Stericycle-MMorris** [comment]
No, sorry I probably wasn't clear.

1) Repo with text file in it
2) Edit the text file, add a character - then commit with a title 'Added missing publish.xml files for DriveAuto' and message Resolving OPS-1975 incorrect DB name in DACPAC variables'
3) Commit to repo
4) Azure DevOps - Run a build that includes these two commits - includes the build notes generator - get the error (No release in Azure just build)

**Comment by Stericycle-MMorris** [comment]
My releasetemplate.md just uses buildDetails because its only a build not a release:
(I just build in Azure DevOps and then pass the artifacts to Octopus Cloud for it to do releases)

# Release notes 
**Build Number**  : ${buildDetails.buildNumber} 
**Build started** : ${buildDetails.startTime}  
**Source Branch** : ${buildDetails.sourceBranch}  

### Associated work items  
@@WILOOP@@  
* ** ${widetail.fields['System.WorkItemType']} ${widetail.id} ** Assigned by: ${widetail.fields['System.AssignedTo']}  ${widetail.fields['System.Title']}  
@@WILOOP@@  

### Associated commits
@@CSLOOP@@  
* ** ID ${csdetail.id} ** Commited by:  ${csdetail.author.displayName} (${csdetail.author.uniqueName}) ${csdetail.message}
@@CSLOOP@@

**Comment by rfennell** [comment]
Sorry still struggling I'm being slow today

> 2. Edit the text file, add a character - then commit with a title 'Added missing publish.xml files for DriveAuto' and message Resolving OPS-1975 incorrect DB name in DACPAC variables'

What do you mean by title and message ? 

- A commit message the title on the first lines, a newlines and then the message - as done in VS Code of Visual Studio
- Some git command line format

`git commit -m "the subject" -m "the first line"$'\n'"the second line"$'\n'"the third line" `

Also 

> 3. Commit to repo

Can this 2nd commit be anything or do you mean push to Azure DevOps

**Comment by Stericycle-MMorris** [comment]
Sorry, I tend to use tools, in this case GitHub desktop
![image](https://user-images.githubusercontent.com/42348904/73955749-c0642080-48fb-11ea-8d37-0706b4dd9404.png)

So:
Summary = 'Added missing publish.xml files for DriveAuto' 
Description = 'Resolving OPS-1975 incorrect DB name in DACPAC variables'

I'm not entirely sure what command that's doing in the background.

Commit to repo = push the above changes to Github and thus trigger the Azure DevOps CI (or be available for a manual build)

**Comment by rfennell** [comment]
OK that make sense. i will work out that that client does. I bet it just adds the newline between the summary and description

**Comment by rfennell** [comment]
The issue seems to be with the creds passed into the function that expands truncated commits. 

I have spent a couple of hours trying to get a repo, and can't. I can't seem to create a commit that is truncated. I used your text and GitHub Desktop.

Until I manage a repo not sure there is much I can do.

**Comment by Stericycle-MMorris** [comment]
I'll see if I cannot make a public repo that has the issue although if it's authentication issue then it probably won't replicate what my private repo's are doing.

I'll try to get a better replication over the weekend and see if I can pin down the issue more specifically.

**Comment by Stericycle-MMorris** [comment]
https://github.com/ExpertSolutionsUK/DevOpsTest

Is a public repo that reproduces this issue. 

It has:
YAML file for DevOps
ReleaseTemplate.md in documentation folder
Readme.md to modify to create new commits

You need an Azure DevOps pipeline based off the YAML file (just release notes task), for the last test builds I literately just used the default DevOps project (private), disabled all but pipelines and linked to the repo / YAML file.

**Comment by rfennell** [comment]
I have repo.

1. If I fork your repo in Github, add a commit (with the failing title and description) and run a build it fails as you have seen
2. However if I import the repo into Azure DevOps and run the build it seems to work.

**So can I confirm you are building directly from A GitHub hosted repo?**

At this point it is worth saying is extension was only ever designed and tested to run against Azure DevOps hosted repos. There is an enhancement request to add full support for GitHub issues, but has not been looked into.

I guess this is going to be the first step in enhancing GitHub support.

I will try to get to the bottom of it.

**Comment by Stericycle-MMorris** [comment]
I can confirm that I am building directly from a GitHub hosted repository.

Using the Third-Party (Microsoft) "Azure Pipelines (OAuth)"  service / plugin?

Just a standard GitHub / Azure DevOps setup

**Comment by rfennell** [comment]
I have found the cause problem and and released potentially a fix as 2.21.1.

The issue was that in some cases the commit message is truncated (to the title). If this is detected an extra call is made to get the rest of the message. The problem is that this extra call is made using the RestClient used for the other Azure Devops API calls i.e. with the Azure DevOps creds. The problem was in your case that as the repo was on GitHub so these were the wrong credentials.

The fix I have done will work for public GitHub repos (as no creds are passed). However, I fear in your case this won't work if the repo is not public.

I see two options
1. Get the creds from the connector (not sure how to do this or tie the call to the right GitHub connector if there are many - this needs thought)
1. Put in the option to not expand truncated commit messages, at least fail in a nicer manner

Any thoughts?

**Comment by rfennell** [comment]
Just checked it on a private repo, and as suspected it fails to expand the commit message, but does fail gracefully - you at least get the title

**Comment by Stericycle-MMorris** [comment]
Well in my use case (and hence why I didn't spot this immediately) most commits have just a title and further information is just a clarification or note. Typically a title has a issue or ticket number in it which is used for storing most of the info. Therefore it failing gracefully is acceptable and you can note in the documentation that GitHub will have truncated commit messages for future users.

This ticket was fixing credentials error which appears to have been done and fully adding GitHub commit / credentials seems like a future todo feature. I'd stick it on the todo list until you have a chance to look at the Azure/GitHub Connector.

**Comment by Stericycle-MMorris** [comment]
Also I confirm that my test repo / build project is no longer erroring.

**Comment by rfennell** [comment]
@Stericycle-MMorris  glad I got yo unblocked

I am not going to close this, as I think the thread is useful. Instead I am going to add the enhancement label. I have also, as you suggested updated the WIKI

**Comment by coonmoo** [comment]
We are also experiencing the issue in Azure DevOps, but we are using Azure Repositories.

```
2020-02-28T15:30:10.0241448Z Task         : Generate Release Notes (Crossplatform)
2020-02-28T15:30:10.0241879Z Description  : Generates a release notes file in a format of your choice from the build or release history
2020-02-28T15:30:10.0242261Z Version      : 2.22.1
2020-02-28T15:30:10.0242515Z Author       : Black Marble
2020-02-28T15:30:10.0243588Z Help         : Version: 2.22.1. [More Information](https://github.com/rfennell/vNextBuild/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task/)
.
.
.
2020-02-28T15:30:12.4014765Z Detected 175 commits/changesets and 17 workitems between the builds.
2020-02-28T15:30:12.4017292Z 
2020-02-28T15:30:12.4045815Z Expanding the truncated commit messages...
2020-02-28T15:30:13.3999857Z ##[error]Error: Failed request: (401) 
2020-02-28T15:30:13.4107302Z ##[error]Error: Failed request: (401) 
2020-02-28T15:30:13.4111111Z ##[section]Finishing: Generate Release Notes
```

We recently migrated from Bitbucket to DevOps, maybe it has something to do with that?
Is there a way to disable expanding?

**Comment by rfennell** [comment]
@coonmoo sorry I missed you comment, are you still suffering the same issue?

I can't see why migrating from Bitbucket should cause an issue, all the information is now in Azure DevOps. But 401 does suggest an access issue.  

In answer to your question currently there is no means to switch off expanding message, but it should fails safe and continue if gets an error

**Comment by rfennell** [comment]
Expanding commit messages (showing not just the first line) has been provided since 2.21.1. However there was a limitation. The expand worked for Azure DevOps hosted repo both public and private, and also for public GitHub hosted repos but fails for private GitHub repos.

I have done an extensive investigation to try to get around this problem. But there are two issues

- In a build there is a GitHub Service Connection that allows the cloning of the repo. The problem is this does not expose a access token
- In a release there is no even this Service Connection.

The issues therefore is that the token is just not available. The only option is therefore to allow a [GitHub PAT](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) to be provided to this task as parameter `gitHubPat`

This new features as been released as 2.43.2
